steps/base.py

```diff
@@ -176,35 +176,69 @@ def _cached_transform(self, step_inputs):
     def adapt(self, step_inputs):
         logger.info('step {} adapting inputs'.format(self.name))
         adapted_steps = {}
-        for adapted_name, mapping in self.adapter.items():
-            if isinstance(mapping, str):
-                adapted_steps[adapted_name] = step_inputs[mapping]
-            else:
-                if len(mapping) == 2:
-                    (step_mapping, func) = mapping
-                elif len(mapping) == 1:
-                    step_mapping = mapping
-                    func = take_first_inputs
-                else:
-                    raise ValueError('wrong mapping specified')
-
-                raw_inputs = [step_inputs[step_name][step_var] for step_name, step_var in step_mapping]
-                adapted_steps[adapted_name] = func(raw_inputs)
+        for adapted_name, recipe in self.adapter.items():
+            try:
+                adapted_steps[adapted_name] = self._adapt_one_name(step_inputs, recipe)
+            except (KeyError, ValueError) as ex:
+                raise StepsError("Error while adapting '{}'".format(adapted_name)) from ex
+
         return adapted_steps
 
     def unpack(self, step_inputs):
         logger.info('step {} unpacking inputs'.format(self.name))
         unpacked_steps = {}
+        key_to_step_names = defaultdict(list)
         for step_name, step_dict in step_inputs.items():
-            unpacked_steps = {**unpacked_steps, **step_dict}
-        return unpacked_steps
+            unpacked_steps.update(step_dict)
+            for key in step_dict.keys():
+                key_to_step_names[key].append(step_name)
```

Comment:
@grzes314 I have to say I use key_to_step_names.setdefault(key, []).append(step_name) instead of defaultdict(list) but it looks way cleaner here.